text/0041-serialization-primitives/0041-serialization-primitives.md

```diff
@@ -0,0 +1,42 @@
++ Feature name: `serialization-primitives`
++ Start date: 2021-11-22
++ RFC PR: [iotaledger/protocol-rfcs#0041](https://github.com/iotaledger/protocol-rfcs/pull/0041)
+
+# Summary
+
+This document introduces the primitives and concepts that are used throughout the IOTA protocol RFCs to describe the binary serialization of objects.
+
+# Motivation
+
+Prior to this document, each RFC contained its own section and version describing the serialization of its objects. This RFC introduces consistent serialization concepts and avoids duplication in other RFCs.
+
+# Detailed design
+## Schemas
+
+Serializable objects are represented by a _schema_. Each schema consists of a list of _fields_, which each have a name and a type. The type of a field can either be a simple data type or another schema, then called subschema.
+
+### Data types
+
+All the supported data types are described in the following table:
+
+| Name         | Description                                                                   |
+| ------------ | ----------------------------------------------------------------------------- |
+| uint8        | An unsigned 8-bit integer encoded in Little Endian.                           |
+| uint16       | An unsigned 16-bit integer encoded in Little Endian.                          |
+| uint32       | An unsigned 32-bit integer encoded in Little Endian.                          |
+| uint64       | An unsigned 64-bit integer encoded in Little Endian.                          |
+| uint256      | An unsigned 256 bits integer encoded in Little Endian.                        |
+| ByteArray[N] | A static size byte array of N bytes.                                          |
+| ByteArray    | A dynamically sized byte array. A leading uint32 denotes its length in bytes. |
```

Comment:
I am not sure this adds clarity. We've introduced the dynamic byte array as a primitive, so that it can be reused in various RFCs and this was also the motivation for using the fixed type of uint32. If this is no longer fixed and can vary, I would rather remove the dynamic byte array completely. In RFCs one would then introduce a new separate field `Length` and the array then has the type `ByteArray[Length]`. This is also how it is already done in the indexation payload in the message RFC.